src/arctica-greeter.vala

```diff
@@ -257,8 +257,17 @@ public class ArcticaGreeter : Object
         sessions.append ("cinnamon");
         sessions.append ("lomiri");
 
+        var excluded_sessions = AGSettings.get_strv (AGSettings.KEY_EXCLUDED_SESSIONS);
+        var includeonly_sessions = AGSettings.get_strv (AGSettings.KEY_INCLUDEONLY_SESSIONS);
+
         if (!AGSettings.get_boolean (AGSettings.KEY_HIDE_WAYLAND_SESSIONS)) {
             foreach (string session in sessions) {
+                if ((includeonly_sessions.length > 0) && (!(session in includeonly_sessions))) {
```

Comment:
I'd change this to

```vala
                if (includeonly_sessions.length > 0) {
                    if (!(session in includeonly_sessions)) {
                        continue;
                    }
                }
                else if (session in excluded_sessions) {
                    continue;
                }
```

This way, we only need to check the include list length once in both cases (which is fast) and especially don't need to go through the whole list again for the excluded list. It'll probably execute a lot faster if vala is caching the array length in a sensible way.

Reply:
Ack. Please adjust.

Reply:
Wait stop! That is not the same algorithm!!!

Reply:
With your algorithm, we only get to the "if (session in excluded_sessions)" clause, if includeonly_sessions is empty. That is not the idea.

Although... you are probably right, there will be no use case for "only include 'dwm', 'mate' and 'plasma', but exclude 'dwm'". This wouldn't make much sense.

Sorry, for extra noise... continue with your proposal...

Reply:
Maybe we should add a comment to the gsettings xml file's description lines... Can you do that?

Reply:
exclude-sessions only takes effect if includeonly-sessions is unused. (summin' like this...)

Reply:
I've been thinking about this for a while and came to the conclusion that both code fragments should lead to the same result:

  - SKIP if included list is non-empty and session is not included there
  - otherwise, the included list is either empty or the session is included in the included list:
    - if included list is empty, we'll have to check the excluded list and SKIP if it's included there
    - otherwise, we don't have to check the excluded list at all because the included list takes precedence (this is what `if ((session in excluded_sessions) && (!(session in includeonly_sessions))) {` tried to achieve) and we know that the session is included there.

Reply:
yes, I agree after a third round of thinking...